tomb-common/src/crypto/rsa.rs

```diff
@@ -0,0 +1,271 @@
+        Ok(fingerprint)
+    }
+
+    /// Writes the public key to a SPKI PEM file.
+    /// # Arguments
+    /// path - The path to the file to write to.
+    pub fn to_pem_file(&self, path: impl AsRef<std::path::Path>) -> Result<()> {
+        self.0
+            .write_public_key_pem_file(path, LineEnding::LF)
+            .map_err(|e| anyhow!(RsaError::ExportToPemFileFailed(anyhow!(e))))
+    }
+
+    /// Reads the public key from a SPKI PEM file.
+    /// # Arguments
+    /// path - The path to the file to read from.
+    pub fn from_pem_file(path: impl AsRef<std::path::Path>) -> Result<Self> {
+        let key = rsa::RsaPublicKey::read_public_key_pem_file(path)?;
+        Ok(Self(key))
+    }
+
+    /// Read the public key from DER bytes.
+    /// # Arguments
+    /// bytes - The DER bytes to read from.
+    pub fn from_der(bytes: &[u8]) -> Result<Self> {
+        let key = rsa::RsaPublicKey::from_public_key_der(bytes)?;
+        Ok(Self(key))
+    }
+}
+
+// #[cfg(test)]
+impl RsaPrivateKey {
+    /// Constructs a new 2048-bit RSA private key.
+    pub fn new() -> Result<Self> {
+        Ok(Self(rsa::RsaPrivateKey::new(
+            &mut rand_core::OsRng,
+            RSA_KEY_SIZE,
+        )?))
+    }
+
+    /// Writes the private key to a PKCS#8 PEM file.
+    ///
+    /// # Arguments
+    /// path - The path to the file to write to.
+    pub fn to_pem_file(&self, path: impl AsRef<std::path::Path>) -> Result<()> {
+        self.0
+            .write_pkcs8_pem_file(path, LineEnding::LF)
+            .map_err(|e| anyhow!(RsaError::ExportToPemFileFailed(anyhow!(e))))
+    }
+
+    /// Reads the private key from a PKCS#8 PEM file.
+    ///
+    /// # Arguments
+    /// path - The path to the file to read from.
+    pub fn from_pem_file(path: impl AsRef<std::path::Path>) -> Result<Self> {
+        let key = rsa::RsaPrivateKey::read_pkcs8_pem_file(path)?;
+        Ok(Self(key))
+    }
+
+    /// Reads the private key from DER bytes.
+    ///
+    /// # Arguments
+    /// bytes - The DER bytes to read from.
+    pub fn from_der(bytes: &[u8]) -> Result<Self> {
+        let key = rsa::RsaPrivateKey::from_pkcs8_der(bytes)?;
+        Ok(Self(key))
+    }
+
+    /// Gets the public key.
+    pub fn get_public_key(&self) -> RsaPublicKey {
+        RsaPublicKey(self.0.to_public_key())
+    }
+}
+
+// #[cfg(test)]
+#[async_trait(?Send)]
+impl ExchangeKey for RsaPublicKey {
+    async fn encrypt(&self, data: &[u8]) -> Result<Vec<u8>> {
+        let padding = Oaep::new::<Sha256>();
+        self.0
+            .encrypt(&mut rand_core::OsRng, padding, data)
+            .map_err(|e| anyhow!(RsaError::EncryptionFailed(anyhow!(e))))
+    }
+
+    async fn from_modulus(modulus: &[u8]) -> Result<Self> {
+        let n = BigUint::from_bytes_le(modulus);
+        let e = BigUint::from(PUBLIC_KEY_EXPONENT);
+
+        Ok(Self(
+            rsa::RsaPublicKey::new(n, e).map_err(|e| RsaError::InvalidPublicKey(anyhow!(e)))?,
+        ))
+    }
+}
```

Comment:
Does this actually get used anywhere other than in the test for it? If not we probably don't want it.

Reply:
This is required for implementing `ExchangKey` 🤷‍♂️